manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml

```diff
@@ -565,6 +565,12 @@ data:
           {{ toYaml $value | indent 4 }}
           {{ end }}
           {{- end }}
+      {{- if .ProxyConfig.ProxyMetadata.ISTIO_META_DNS_CAPTURE }}
+      dnsConfig:
+        options:
+        - name: "ndots"
+          value: "4"
```

Comment:
can you give more explanation about this change? also how does this work if the user already set dnsConfig?

Reply:
If the user set this config already, then i think it gets overwritten. That is the reason why I guarded the dns thing with an if block. The ndots thing says that if a given hostname has atleast this many dots, then try to resolve the hostname as a FQDN first before attempting to expand the hostname with all the dns search namespaces. Its a standard option in resolv.conf. 
K8S increased ndots to 5 which results in dns slow downs in some cases. But in our case, due to slowdown, lots of tests start failing due to timeouts on resolution. https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html . 
In a subsequent PR, we should make this configurable.

Reply:
I don't think we should mix us resolving DNS with a performance optimization. If we need this optimization to have the DNS resolution working than the performance of the envoy DNS stuff must be so bad that it's not viable in it's current state?

Reply:
Note this will break existing application as well as it can change the behavior

Reply:
It wont break anything actually. All this does is force the resolution of the application supplied hostname as is, first, before actually trying out all the search namespaces in the resolv.conf. But the correctness of the system will never be impacted. So, if the app does http://a, it will still try to be resolved as a.ns.svc.cluster.local or a.ns or a.ns.google.internal or so on.. same goes for http://a.ns, or http://a.ns.svc, when it does http://a.ns.svc.cluster.local, then instead of trying to resolve a.ns.svc.cluster.local.ns.svc.cluster.local, the system will first try to resolve a.ns.svc.cluster.local as is first before attempting to add other prefixes.

Now, the reason I had to do this hack is because of the following inherent problem with intercepting DNS. Without an explicit nameserver, the app container will try to resolve every dns name (containing < ndots) with all the N suffixes in the resolv.conf. When it reaches Envoy, since one more of these names are not really in envoy's dns list (e.g., a.ns.svc.cluster.local.svc.cluster.local) envoy will try to resolve it using the resolv.conf in its container - where once again, it goes through all the N suffixies, i.e. envoy tries a.ns.svc.cluster.local.svc.cluster.local.svc.cluster.local. So in effect, it becomes a O(N2) thing which causes lots of our tests to timeout waiting for dns resolution. By setting ndots to 4, we reduce the number of useless lookups to a smaller figure. Its not the best way to fix the problem but its a start. The ideal solution is to make sure envoy (or istiod) does not search through the DNS search namespaces but rather only resolves the hosts provided as is.